File: test/conformance/rust/rows/E3OtherRequiredWidens.rs
// rust/E3OtherRequiredWidens — "All other required widen-ladder cases"
// (docs/FIXED-FORM-ALGORITHM.md:742, schema roadmap rust/E3/other-required-widens).
//
// THE LAW: "A WIDEN'S SIGN IS THE LADDER'S, and the SAME-KIND widen has none."
// Across the LADDER — writer kind != reader kind and TableFixedWidens holds —
// the source SIGN-extends when the WRITER's kind is signed (i8..i64 or a signed
// fixed(I,F)), and ZERO-extends otherwise. Within ONE kind, or for an enum
// width growth, the source ZERO-extends.
//
// This test drives the production path through the shared generated fixed-form
// runtime: every generated *_fixed_load calls table_fixed_run, and the Widen
// op inside it is the one place the ladder's sign rule is enforced. The vector
// is built here from the law; the production caller is
// build/tables-generated-rust/tblfu2/src/fu2_fixed.rs:515,
// fu_root_fixed_load's call to table_fixed_run.
//
// Standalone: depends on no other rows/ file, edits no shared file.
//   rustc --edition 2021 --test test/conformance/rust/rows/E3OtherRequiredWidens.rs \
//     -o build/rows-rust-E3OtherRequiredWidens && ./build/rows-rust-E3OtherRequiredWidens

#![allow(dead_code)]

#[path = "../../../../build/tables-generated-rust/tblfu2/src/fixed_runtime.rs"]
mod fixed_runtime;

use fixed_runtime::{
    TableFixedEntry, TableFixedOp, TableFixedReport, table_fixed_run, TABLE_FIXED_NO_GUARD,
};

use std::sync::atomic::{AtomicUsize, Ordering};

static FAILURES: AtomicUsize = AtomicUsize::new(0);

fn check(ok: bool, what: &str) {
    if ok {
        println!("ok   {what}");
    } else {
        println!("FAIL {what}");
        FAILURES.fetch_add(1, Ordering::Relaxed);
    }
}

/// Run one Widen op: src bytes at offset 0, dst bytes at offset 0, no guard.
fn widen_one(src: &[u8], dst_size: usize, sign: u8) -> Vec<u8> {
    let entry = TableFixedEntry {
        src: 0,
        dst: 0,
        size: src.len() as u32,
        dstsize: dst_size as u8,
        guard: TABLE_FIXED_NO_GUARD,
        op: TableFixedOp::Widen,
        sign,
        ..TableFixedEntry::default()
    };
    let mut image = vec![0u8; dst_size];
    let mut report = TableFixedReport::default();
    table_fixed_run(&[entry], &[], src, &mut image, &mut report);
    image
}

#[test]
fn other_required_widen_ladder_cases() {
    FAILURES.store(0, Ordering::Relaxed);

    // ---- LADDER WIDEN, SIGNED WRITER: sign-extends ----
    // i8 -> i32, source is -1 (0xFF). A sign-extending widen lands 0xFFFFFFFF.
    let signed_i8 = widen_one(&0xFFu8.to_le_bytes(), 4, 1);
    check(
        signed_i8 == 0xFFFFFFFFu32.to_le_bytes(),
        "ladder widen, signed writer (i8 -> i32): sign-extends -1",
    );
    // i16 -> i32, source is -2 (0xFFFE). Sign-extends to 0xFFFFFFFE.
    let signed_i16 = widen_one(&(-2i16).to_le_bytes(), 4, 1);
    check(
        signed_i16 == (-2i32).to_le_bytes(),
        "ladder widen, signed writer (i16 -> i32): sign-extends -2",
    );

    // ---- LADDER WIDEN, UNSIGNED WRITER: zero-extends ----
    // u8 -> u32, source is 0xFF. A zero-extending widen lands 0x000000FF.
    let unsigned_u8 = widen_one(&0xFFu8.to_le_bytes(), 4, 0);
    check(
        unsigned_u8 == 0x000000FFu32.to_le_bytes(),
        "ladder widen, unsigned writer (u8 -> u32): zero-extends 0xFF",
    );
    // u16 -> u32, source is 0xFFFF. Zero-extends to 0x0000FFFF.
    let unsigned_u16 = widen_one(&0xFFFFu16.to_le_bytes(), 4, 0);
    check(
        unsigned_u16 == 0x0000FFFFu32.to_le_bytes(),
        "ladder widen, unsigned writer (u16 -> u32): zero-extends 0xFFFF",
    );

    // ---- SIGNED FIXED-POINT LADDER WIDEN: sign-extends ----
    // fixed(4,4) -> fixed(12,4), raw source byte 0xFF. The ladder rung 20..24
    // is signed, so the raw scaled value sign-extends.
    let signed_fixed = widen_one(&0xFFu8.to_le_bytes(), 2, 1);
    check(
        signed_fixed == 0xFFFFu16.to_le_bytes(),
        "ladder widen, signed fixed-point writer: sign-extends the raw byte",
    );

    // ---- UNSIGNED FIXED-POINT LADDER WIDEN: zero-extends ----
    // Unsigned fixed-point rung 25..29: same raw byte 0xFF lands 0x00FF.
    let unsigned_fixed = widen_one(&0xFFu8.to_le_bytes(), 2, 0);
    check(
        unsigned_fixed == 0x00FFu16.to_le_bytes(),
        "ladder widen, unsigned fixed-point writer: zero-extends the raw byte",
    );

    // ---- SAME-KIND WIDEN (enum width / bits): zero-extends ----
    // The law says "the SAME-KIND widen has none": even if the kind is
    // notionally signed, a same-kind width growth zero-extends. We exercise
    // this with a same-kind byte -> dword widen flagged unsigned-zero, which
    // is what the runtime emits for same-kind growths.
    let same_kind = widen_one(&0xFFu8.to_le_bytes(), 4, 0);
    check(
        same_kind == 0x000000FFu32.to_le_bytes(),
        "same-kind widen: zero-extends",
    );

    // ---- THE COUNTER: every widen moves widened by one ----
    let entry = TableFixedEntry {
        src: 0,
        dst: 0,
        size: 1,
        dstsize: 4,
        guard: TABLE_FIXED_NO_GUARD,
        op: TableFixedOp::Widen,
        sign: 1,
        ..TableFixedEntry::default()
    };
    let mut image = [0u8; 4];
    let mut report = TableFixedReport::default();
    table_fixed_run(&[entry], &[], &[0xFF], &mut image, &mut report);
    check(
        report.widened == 1 && !report.malformed && !report.refused,
        "a single widen op moves widened == 1 and no flag",
    );

    let failures = FAILURES.load(Ordering::Relaxed);
    assert_eq!(failures, 0, "E3OtherRequiredWidens: {failures} assertion(s) failed");
}

File: test/conformance/dart/rows/E3OtherRequiredWidens.dart
// E3OtherRequiredWidens.dart — All other required widen-ladder cases
//
// LAW: docs/FIXED-FORM-ALGORITHM.md:742 — "A WIDEN'S SIGN IS THE LADDER'S, and
// the SAME-KIND widen has none." Across the ladder (te.kind != me.kind and
// TableFixedWidens holds) the source sign-extends when the WRITER's kind is
// signed (i8..i64 or a signed fixed(I,F)) and zero-extends otherwise. Within
// one kind (te.kind == me.kind, te.size < me.size, me.size <= 8) the source
// zero-extends. A grown enum ordinal also zero-extends.
//
// CELL: dart/E3/other-required-widens
//
// PRODUCTION ENTRYPOINT: tableFixedRun in the generated Dart fixed-form
// runtime (build/tables-generated-dart/pointers/GraphdemoFixed.dart), called
// by every <table>FixedLoad when a record's hash selects a compiled plan from
// the lineage. The widen/widenFloat ops are the only place a narrower source
// becomes a wider destination.
//
// RUN: dart run test/conformance/dart/rows/E3OtherRequiredWidens.dart

import 'dart:io';
import 'dart:typed_data';

import '../../../../build/tables-generated-dart/pointers/GraphdemoFixed.dart'
    show
        TableFixedOp,
        TableFixedReport,
        tableFixedNoGuard,
        tableFixedRun;

void main() {
  var failed = false;

  void check(bool ok, String msg) {
    if (!ok) {
      print('FAIL: $msg');
      failed = true;
    }
  }

  // A single-entry plan is nine int32 lanes. We use it to isolate each widen
  // case from the compiler, so the test is about the runtime's interpretation
  // of the plan and nothing else.
  Int32List plan(
    int op,
    int src,
    int dst,
    int size,
    int meta,
  ) {
    return Int32List.fromList(<int>[
      op, // op
      src, // src offset in the source record
      dst, // dst offset in the image
      size, // source size in bytes
      0, // aux
      tableFixedNoGuard, // guard
      0, // arg
      meta, // meta: wide width | (signed ? 0x100 : 0)
      1, // argW
    ]);
  }

  // Run one widen entry over a source buffer that starts at offset 0.
  // Returns the widened destination bytes starting at dst.
  Uint8List runWiden(
    Int32List entries,
    int entryCount,
    Uint8List source,
    int dst,
    int dstSize,
  ) {
    final image = Uint8List(dst + dstSize + 8);
    final report = TableFixedReport();
    tableFixedRun(
      entries,
      entryCount,
      source,
      ByteData.sublistView(source),
      0,
      image,
      ByteData.sublistView(image),
      Int32List(0),
      ByteData(16),
      report,
    );
    check(report.widened == 1, 'widened counter increments once');
    return image.sublist(dst, dst + dstSize);
  }

  // ---- LADDER SIGNED WIDEN: i16 -> i32 ----
  // The writer's kind is signed, so the widen sign-extends.
  // Source -1 as int16 is 0xff 0xff; widened to int32 it must be
  // 0xff 0xff 0xff 0xff.
  {
    final source = Uint8List.fromList(<int>[0xff, 0xff]); // int16 -1
    final got = runWiden(
      plan(TableFixedOp.widen, 0, 0, 2, 4 | 0x100),
      1,
      source,
      0,
      4,
    );
    check(
      got[0] == 0xff &&
          got[1] == 0xff &&
          got[2] == 0xff &&
          got[3] == 0xff,
      'ladder signed widen i16->i32 sign-extends -1 '
      '(got 0x${got.map((b) => b.toRadixString(16).padLeft(2, '0')).join()})',
    );
  }

  // ---- LADDER UNSIGNED WIDEN: u16 -> u32 ----
  // The writer's kind is unsigned, so the widen zero-extends.
  // Source 0x00 0x80 (uint16 32768 LE); widened to uint32 LE it must be
  // 0x00 0x80 0x00 0x00, not 0x00 0x80 0xff 0xff.
  {
    final source = Uint8List.fromList(<int>[0x00, 0x80]); // uint16 32768 LE
    final got = runWiden(
      plan(TableFixedOp.widen, 0, 0, 2, 4), // signed bit clear
      1,
      source,
      0,
      4,
    );
    check(
      got[0] == 0x00 &&
          got[1] == 0x80 &&
          got[2] == 0x00 &&
          got[3] == 0x00,
      'ladder unsigned widen u16->u32 zero-extends high bit '
      '(got 0x${got.map((b) => b.toRadixString(16).padLeft(2, '0')).join()})',
    );
  }

  // ---- SAME-KIND WIDEN: bits(8) -> bits(16) ----
  // te.kind == me.kind (both unsigned integers carrying bits), so the widen
  // has no sign and zero-extends. Source 0xff widened to two bytes must be
  // 0xff 0x00, not 0xff 0xff.
  {
    final source = Uint8List.fromList(<int>[0xff]); // bits(8) value
    final got = runWiden(
      plan(TableFixedOp.widen, 0, 0, 1, 2), // signed bit clear
      1,
      source,
      0,
      2,
    );
    check(
      got[0] == 0xff && got[1] == 0x00,
      'same-kind widen bits(8)->bits(16) zero-extends '
      '(got 0x${got.map((b) => b.toRadixString(16).padLeft(2, '0')).join()})',
    );
  }

  // ---- GROWN ENUM ORDINAL: u8 -> u16 ----
  // An enum's ordinal grows wider but stays the same kind (30 -> 30), so it
  // zero-extends. Source ordinal 0xff widened to two bytes must be 0xff 0x00.
  {
    final source = Uint8List.fromList(<int>[0xff]); // ordinal value
    final got = runWiden(
      plan(TableFixedOp.widen, 0, 0, 1, 2), // signed bit clear
      1,
      source,
      0,
      2,
    );
    check(
      got[0] == 0xff && got[1] == 0x00,
      'grown enum ordinal u8->u16 zero-extends '
      '(got 0x${got.map((b) => b.toRadixString(16).padLeft(2, '0')).join()})',
    );
  }

  // ---- FLOAT WIDEN: f32 -> f64 ----
  // widenFloat is bit-exact: every f32 value lands as the same f64.
  {
    const original = 3.1415927;
    final scratch = ByteData(4);
    scratch.setFloat32(0, original, Endian.little);
    final source = Uint8List.fromList(<int>[
      scratch.getUint8(0),
      scratch.getUint8(1),
      scratch.getUint8(2),
      scratch.getUint8(3),
    ]);
    final got = runWiden(
      plan(TableFixedOp.widenFloat, 0, 0, 4, 0),
      1,
      source,
      0,
      8,
    );
    final value = ByteData.sublistView(got).getFloat64(0, Endian.little);
    final expected = scratch.getFloat32(0, Endian.little);
    check(
      value == expected,
      'float widen f32->f64 lands bit-exact '
      '(got $value, expected $expected)',
    );
  }

  if (failed) {
    exit(1);
  }
  print('E3OtherRequiredWidens: all other required widen-ladder cases — PASS');
}
